include/deal.II/base/table_indices.h

```diff
@@ -43,11 +45,71 @@ class TableIndices
 public:
 
   /**
-   * Default constructor. It sets all indices to zero.
+   * Default constructor. This constructor sets all indices to zero.
    */
   TableIndices();
 
   /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<1> and initializes the single
+   * index with @p index0.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from one.
+   */
+  explicit TableIndices (const unsigned int index0);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<2> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from two.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<3> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from three.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<4> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from four.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3);
+
+  /**
+   * Constructor. This is the appropriate constructor for an
+   * object of type TableIndices<5> and initializes the
+   * indices stored by this object by the given arguments.
+   *
+   * This constructor will result in a compiler error if
+   * the template argument @p N is different from five.
+   */
+  TableIndices (const unsigned int index0,
+                const unsigned int index1,
+                const unsigned int index2,
+                const unsigned int index3,
+                const unsigned int index4);
```

Comment:
@masterleinad -- yes, I thought about this approach. The problem is that you will get an error about "no such constructor" when you write
```
  TableIndices<2> indices(2,3);
```
That's particularly unreadable because the arguments seem totally sensible, but they happen to be signed integers, not unsigned integers as we want. The `enable_if` idiom is really not very useful here because the compiler just says that there is no matching function -- but it doesn't say *why* the constructors of the class don't match.

This may, however, be useful as a `static_assert`. I considered this, but it's not quite clear to me what I'm supposed to test there. In essence, I want to test "given type could be converted to an `unsigned int` in a function argument list without warning". The important part is the last half sentence -- that's because when you call a function that takes an `unsigned int` with `1.234`, then you'll get a *warning* (rightfully so!), but the code is valid in itself -- and so you *won't* get a warning if the conversion is only tested in a `static_assert`.

What I would love to have is something like
```
  TableIndices (const unsigned int... args) {
    static_assert(sizeof...(args) == N);
    values = {args...};
  }
```
but this idiom does not exist in C++ as far as I know :-(